tests/TestCase/AuthenticationServiceTest.php

```diff
@@ -424,6 +424,25 @@ public function testPersistIdentityArray()
         $this->assertInstanceOf(IdentityInterface::class, $result['request']->getAttribute('identity'));
     }
 
+    /**
+     * Test that the persistIdentity() called with an identity instance sets
+     * this instance as a request attribute
+     *
+     * @return void
+     */
+    public function testPersistIdentityInstance()
+    {
+        $request = new ServerRequest();
+        $response = new Response();
+        $identity = new Identity([]);
+
+        $service = new AuthenticationService();
+
+        $result = $service->persistIdentity($request, $response, $identity);
```

Comment:
Passing an `Identity` instance here would mean `SessionAuthenticate` would store that instance in the session. But when one uses form login the instance stored in session is an entity, not an `IdentityInterface` instance.

IMO it's better to avoid such discrepancy.

Reply:
That said a clean way to fix the regression would be make `buildIdentity()` simply return the argument if it's already an `IdentityInterface` instance. That's would also avoid the object type checks at multiple places.

Reply:
> The authenticators should always deal with the "raw" identity object and decorating it with IdentityInterface should be left to AuthenticationService.

The thing is that sometimes the "raw" identity object is an identity instance itself. If your User entity already implements `IdentityInterface` that is. And this test covers that case.

Reply:
> If your User entity already implements IdentityInterface that is.

Ah, I had overlooked that case.

Reply:
@ADmad done and done.